src/porepy/compositional/compositional_mixins.py

```diff
@@ -0,0 +1,1087 @@
+
+            dependencies += independent_overall_fractions
+
+        # casting to include mortar greeds and variables as return types.
+        # This is used as an argument for Surrogate factories, so we must have the
+        # mortars as well
+        return cast(
+            Sequence[Callable[[pp.GridLikeSequence], pp.ad.Variable]], dependencies
+        )
+
+    def fluid_density(self, domains: pp.SubdomainsOrBoundaries) -> pp.ad.Operator:
+        """Returns an operaror by calling phase saturations and densities, and
+        performing a weighed sum."""
+
+        op = pp.ad.sum_operator_list(
+            [
+                phase.saturation(domains) * phase.density(domains)
+                for phase in self.fluid_mixture.phases
+            ],
+            "fluid-mixture-density",
+        )
+        return op
+
+    def fluid_specific_volume(
+        self, domains: pp.SubdomainsOrBoundaries
+    ) -> pp.ad.Operator:
+        """Returns the reciprocal of whatever :meth:`fluid_density` returns."""
+        op = self.fluid_density(domains) ** pp.ad.Scalar(-1)
+        op.set_name("fluid-mixture-specific-volume")
+        return op
+
+    def fluid_specific_enthalpy(
+        self, domains: pp.SubdomainsOrBoundaries
+    ) -> pp.ad.Operator:
+        """Returns an operaror by calling phase fractions and specific enthalpies, and
+        performing a weighed sum.
+
+        Raises:
+            CompositeModellingError: If :attr:`equilibrium_type` is None, and hence no
+                phase fractions were introduced by :class:`CompositionalVariables`.
+                The consistent definition of the specific molar enthalpy of a fluid
+                mixture always depends on phase fractions.
+
+        """
+
+        if self.equilibrium_type is None:
+            raise CompositionalModellingError(
+                "Attempting to define the (specific) fluid mixture enthalpy as sum of"
+                + " phase enthalpies weighed with phase fractions, even though no"
+                + " equilibrium conditions defined. Per default, phase fractions are"
+                + " only created when an equilibrium formulation is used in the model."
+            )
+
+        op = pp.ad.sum_operator_list(
+            [
+                phase.fraction(domains) * phase.specific_enthalpy(domains)
+                for phase in self.fluid_mixture.phases
+            ],
+            "fluid-mixture-specific-enthalpy",
+        )
+        return op
+
+    def density_of_phase(self, phase: Phase) -> pp.ad.SurrogateFactory:
+        """This base method returns the density of a ``phase`` as a
+        :class:`~porepy.numerics.ad.surrogate_operator.SurrogateFactory`.
+
+        It is populated by the solution strategy, depending on how
+        the equilibrium conditions and constitutive laws are implemented.
+
+        The phase density (like all thermodynamic properties) is a dependent quantity.
+
+        """
+        return pp.ad.SurrogateFactory(
```

Comment:
Do I understand correctly that you use the `SurrogateFactory` to wrap values from the fluid state?

Reply:
Yes, for every thermodynamic property, except volume, which is a reciprocal of density.
I eliminated the volume to safe memory.
Could be the other way round in a volume-based formulation. There it might be better to avoid the ``**(-1)`` operation.